### PR TITLE
router: use new config_impl_test_lib instead of synthetic build target

### PR DESCRIPTION
**Pull request by fcfort**
Signed-off-by: Frank Fort <ffort@google.com>

Description: The build target `//test/common/router:config_impl_test_static` was relying upon the `_lib` target that is created by the `envoy_cc_test` in envoy_build_system.bzl. This target should not be relied upon since it is an implementation detail of the Envoy build system rather than a public build target that we want to allow dependencies upon.

This PR restricts the visibility of these synthetic targets to only the code coverage tools and renames them to end with `_lib_INTERNAL_ONLY` instead. See #6725 for where we started using these synthetic targets in our `BUILD` files.

This PR also hides the `envoy_cc_test_infrastructure_library` build macro by renaming it to `_envoy_cc_test_infrastructure_library` as per https://docs.bazel.build/versions/master/build-ref.html#load. This macro was not being used outside of envoy_build_system.bzl.

Note this does not break coverage builds since coverage pulls in the synthetic targets via the `coverage_test_lib` tag, see [generated_coverage_BUILD.txt](https://github.com/envoyproxy/envoy/files/3163463/generated_coverage_BUILD.txt) for the generated coverage `BUILD` file.

Risk Level: Low
Testing: Ran `bazel test //test/common/...`. and ran `test/run_envoy_bazel_coverage.sh`. 
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by fcfort** [comment]
@lizan @htuch

**Comment by fcfort** [comment]
Looks like there was somewhere else relying upon these _lib targets that I missed let me take a look.

**Comment by lizan** [comment]
I'm not very thrilled with this approach though not a strong opinion, this is the issue how Google internal `envoy_build_system.bzl` is implemented. Also I'm worrying about other repository that may depends on the test libs. We didn't guard them with bazel visibility or document it so I don't think it is really an implementation details which we should hide.

The alternative would be either:
- Fix your internal `envoy_build_system.bzl`
- Exposing `linkstatic` in `envoy_cc_test`.

@htuch thoughts?

File: bazel/envoy_build_system.bzl [hunk 1]
```diff
@@ -456,7 +456,7 @@ def envoy_cc_test(
     if coverage:
         test_lib_tags.append("coverage_test_lib")
     envoy_cc_test_library(
-        name = name + "_lib",
+        name = name + "__lib",
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: I'd suggest that we make this `_lib_internal_only`. That's clear and doesn't have the difficulty in visually distinguishing between one and two underscores :)

Comment (fcfort) [reply bazel/envoy_build_system.bzl]: Done. Went with `_lib_INTERNAL_ONLY`.

**Comment by fcfort** [comment]
> @lizan good points, I think though that anyone relying on this behavior is probably sophisticated enough to fix when this merges. @fcfort can you fix the visibility as well, so that we can prevent folks forming these dependencies?

Done, see updated PR description. Basically:
>This PR restrict the visibility of these synthetic targets to only the code coverage tools and renames them to end with `_lib_INTERNAL_ONLY` instead. See #6725 for where we started using these synthetic targets in our `BUILD` files.

>This PR also hides the `envoy_cc_test_infrastructure_library` build macro by renaming it to `_envoy_cc_test_infrastructure_library` as per https://docs.bazel.build/versions/master/build-ref.html#load. This macro was not being used outside of envoy_build_system.bzl.

**Comment by fcfort** [comment]
As an aside it may be worth fixing some our uses of `envoy_cc_test_library`. By default it declares all targets as `//visibility:public` which will hamper refactoring efforts.

File: bazel/envoy_build_system.bzl [hunk 2]
```diff
@@ -455,15 +455,17 @@ def envoy_cc_test(
     test_lib_tags = []
     if coverage:
         test_lib_tags.append("coverage_test_lib")
-    envoy_cc_test_library(
-        name = name + "_lib",
+    _envoy_cc_test_infrastructure_library(
+        name = name + "_lib_INTERNAL_ONLY",
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: I think `_internal_only` is more consistent with our usual naming and still clear enough.

File: bazel/envoy_build_system.bzl [hunk 3]
```diff
@@ -455,15 +455,17 @@ def envoy_cc_test(
...
         srcs = srcs,
         data = data,
         external_deps = external_deps,
-        deps = deps,
+        deps = deps + [repository + "//test/test_common:printers_includes"],
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Why this change? What do we gain by replacing envoy_cc_test_library?

File: bazel/envoy_build_system.bzl [hunk 4]
```diff
@@ -455,15 +455,17 @@ def envoy_cc_test(
...
         repository = repository,
         tags = test_lib_tags,
         copts = copts,
+        # Restrict only to the code coverage tools.
+        visibility = ["@envoy//test/coverage:__pkg__"],
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: Can you verify that the fuzzers targets needed at https://github.com/google/oss-fuzz/blob/master/projects/envoy/build.sh#L62 still build? I.e. the `_driverless` variants of each fuzz target. This isn't caught by CI.

File: bazel/envoy_build_system.bzl (hunk 3 above)
Comment (fcfort) [reply bazel/envoy_build_system.bzl]: We cannot continue to use `envoy_cc_test_library` here since it declares all targets created with it as `//visibility:public`, which means we cannot restrict the visibility of the `_lib_internal_only` targets to `@envoy//test/coverage:__pkg__`.

File: bazel/envoy_build_system.bzl (hunk 4 above)
Comment (fcfort) [reply bazel/envoy_build_system.bzl]: Done. Verified that `bazel build $(bazel query 'filter("_driverless", kind(".* rule", //...))')` progresses to the linking stage (despite encountering a library error there) and that  `bazel build $(bazel query 'filter("_corpus$", kind(".* rule", //...))')` finishes successfully.

File: bazel/envoy_build_system.bzl (hunk 2 above)
Comment (fcfort) [reply bazel/envoy_build_system.bzl]: Done.
